### PR TITLE
feat(orm): add typed QuerySet retrieval helpers

### LINKED ISSUE #5845: Feature: Add typed QuerySet retrieval helpers
## Prerequisites

- [x] I have searched existing issues to verify this feature hasn't been requested

## Summary

Add typed `in_bulk`, `latest`, `earliest`, and lazy `none` operations to the ORM QuerySet, together with model metadata equivalent to Django's `get_latest_by`.

Split from #5812. Target branch: `develop/0.4.0`.

## Motivation / Problem

Routine keyed retrieval, latest/earliest selection, and composable empty querysets currently require application-specific helpers. The APIs must preserve Reinhardt's generated field guarantees, deterministic results, and lazy execution semantics.

## Proposed Solution

- Add `latest()` and `earliest()` using generated model `get_latest_by` metadata.
- Add `latest_by(...)` and `earliest_by(...)` for one or more generated typed ordering fields.
- Return the ORM not-found error when latest/earliest evaluation finds no row.
- Add `in_bulk` for primary keys and generated `UniqueFieldRef` fields.
- Return `BTreeMap<K, Model>` with `K: Ord`.
- Collapse duplicate requested keys, ignore missing keys, and execute no query for empty input.
- Add a lazy `none()` marker and centralize execution short-circuiting so `all`, `first`, `get`, `count`, `exists`, and future execution paths do not resolve or invoke an executor.
- Provide caller-owned executor variants wherever the equivalent QuerySet execution path already does so.
- Add compile-pass/compile-fail coverage, mock query-count tests, PostgreSQL/MySQL/SQLite tests, and user-facing documentation.

### Durable implementation and verification contracts

- Represent ordering, unique-field lookup, and evaluation SQL with typed `reinhardt-query` statements and expressions rather than unchecked SQL strings in QuerySet code.
- Add compile-pass tests for valid generated typed ordering fields and `UniqueFieldRef` fields, plus compile-fail tests for invalid ordering fields and non-unique fields.
- Add unit tests proving deterministic `BTreeMap` key ordering, duplicate requested-key collapse, and missing-key omission.
- Update the relevant crate README, `lib.rs` module documentation, and ORM user-facing documentation. Include typed-field and caller-owned-executor examples where applicable, state backend limitations explicitly, and document the intentionally stricter-than-Django typed-field and deterministic-map contracts.

## Alternatives Considered

1. Return `HashMap`. This does not provide deterministic iteration order.
2. Return `IndexMap`. This would add a collection contract when sorted key order is sufficient.
3. Accept string field names. This would discard compile-time model and uniqueness checks.

## Is this a breaking change?

No - backward compatible

## Migration Path (if breaking)

Not applicable. These are additive QuerySet capabilities.

## Additional Context

This issue intentionally excludes streaming iteration, query explanation, date projection, row locking, `alias`, `raw`, typed aggregation/annotation, and typed manager upserts.

### PR DESCRIPTION
**Pull request by kent8192**
## Summary

This PR addresses:

- Adds typed `latest`, `earliest`, `latest_by`, and `earliest_by` QuerySet terminals.
- Adds deterministic `in_bulk` and unique-field `in_bulk_by` helpers with caller-owned connection and transaction variants.
- Generates validated `get_latest_by` and unique-field metadata from model declarations.
- Makes `QuerySet::none()` lazy across global and caller-owned terminal operations.
- Documents the typed retrieval API and adds compile-time, mock-executor, and real-backend coverage.

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (code change that neither fixes a bug nor adds a feature)
- [x] Documentation update
- [ ] Performance improvement
- [x] Code quality improvements
- [ ] CI/CD changes
- [ ] Other (please describe):

## Breaking Change Assessment

- [x] **No** - This change is backward compatible
- [ ] **Yes** - This change breaks existing public API or behavior

## Motivation and Context

Typed retrieval helpers eliminate stringly typed ordering and unique-key bulk lookup paths while preserving deterministic results and explicit executor ownership. Lazy empty querysets avoid acquiring a connection when no query can produce rows.

Fixes #5845

Related to: #5845

## How Was This Tested?

- `cargo test --offline -p reinhardt-db --test orm_queryset_builder --test transaction_executor_orm` (104 passed)
- `cargo test --offline -p reinhardt-integration-tests --test orm typed_queryset_retrieval_contract -- --nocapture` (SQLite, PostgreSQL, and MySQL passed)
- `cargo test --offline -p reinhardt-integration-tests --test orm queryset_docs_example_compile_pass` (8 trybuild pass/fail cases passed)
- `cargo fmt --all -- --check`
- `git diff --check`
- `cargo test --offline -p reinhardt-macros model_derive --lib` (57 passed; one pre-existing unrelated ModelForm metadata baseline failure remains)

## Performance Impact

- Empty querysets short-circuit without connection or executor access.
- Bulk retrieval deduplicates keys and issues one typed `IN` query.

## Checklist

- [x] I have followed the [Contributing Guidelines](../blob/main/CONTRIBUTING.md)
- [x] I have followed the [Commit Guidelines](../blob/main/instructions/COMMIT_GUIDELINE.md)
- [x] I have updated the documentation (if applicable)
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing focused unit tests pass locally with my changes
- [x] I have tested with all affected database backends (if applicable)
- [x] I have formatted the code with `cargo fmt --all -- --check`
- [ ] I have checked the code with `cargo make clippy-check`
- [ ] I use self-hosted runner for CI (Repository owner only)

## Related Issues

- Fixes #5845

### Type Label (select one)
- [x] `enhancement` - New feature or improvement
- [ ] `bug` - Bug fix
- [ ] `documentation` - Documentation update
- [ ] `performance` - Performance improvement
- [ ] `refactoring` - Code refactoring
- [ ] `code-quality` - Code quality improvements

### Additional Labels (apply alongside type label when applicable)
- [ ] `breaking-change` - Breaking change

### Scope Label (select all that apply)
- [x] `database` - Database layer, schema, migrations
- [x] `orm` - ORM layer, models, query builder

### Priority Label (for maintainers)
- [x] `high` - Important fix or feature

---

**Additional Context:**

The branch includes an ordinary merge from the latest `develop/0.4.0`; no rebase or history-rewriting push was used.

🤖 Generated with [Codex](https://openai.com/codex)

### CODE REVIEW

File: crates/reinhardt-db/src/orm/expressions.rs [hunk 1]
```diff
@@ -195,6 +265,14 @@ impl<M, T> FieldRef<M, T> {
 		self.name
 	}
 
+	/// Convert this field reference into a type-safe ordering field.
+	pub const fn ordering(&self) -> OrderingField<M> {
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/expressions.rs]: No code change is needed. Model field accessors use their physical db_column, while relation traversal uses a distinct RelatedFieldRef API; a relation field does not provide an ordering proof for an unmapped relation column.\n\n— Codex

File: crates/reinhardt-db/src/orm/query.rs [hunk 2]
```diff
@@ -5946,10 +5965,486 @@ where
...
+		keys: BTreeSet<T::PrimaryKey>,
+	) -> reinhardt_core::exception::Result<BTreeMap<T::PrimaryKey, T>>
+	where
+		T: serde::de::DeserializeOwned,
+		T::PrimaryKey: DatabaseField + Ord,
+		E: OrmExecutor,
+	{
+		let filter = super::expressions::FieldRef::<T, T::PrimaryKey>::new(T::primary_key_column())
+			.is_in(keys);
+		let rows = self.clone().filter(filter).all_with_db(conn).await?;
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs]: Implemented in 8ed7ba3cf1. Bulk retrieval now splits key sets below SQLite's bind limit (and uses each server backend's larger limit), then merges each batch into the deterministic BTreeMap result.\n\n— Codex

File: crates/reinhardt-db/tests/transaction_executor_orm.rs [hunk 3]
```diff
@@ -538,6 +637,340 @@ async fn dyn_orm_executor_saves_new_and_existing_articles() {
...
+			.unwrap()
+			.id,
+		Some(2)
+	);
+	assert_eq!(executor.calls.len(), 4);
+	assert!(
+		executor.calls[0]
+			.sql
+			.contains("ORDER BY \"article_id\" DESC")
+	);
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/tests/transaction_executor_orm.rs]: Implemented in 8ed7ba3cf1. The retrieval SQL assertions now compare the complete SQL strings and parameter vectors.\n\n— Codex

File: crates/reinhardt-db/src/orm/query.rs [hunk 4]
```diff
@@ -5946,10 +5965,486 @@ where
...
+		conn: &mut E,
+		ordering: Vec<String>,
+	) -> reinhardt_core::exception::Result<T>
+	where
+		T: serde::de::DeserializeOwned,
+		E: OrmExecutor,
+	{
+		let mut queryset = self.clone();
+		queryset.order_by_fields = ordering;
+		queryset.limit = Some(1);
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:7316]: Implemented in 8ed7ba3cf1. All single-row retrieval paths now reject a queryset that already has a limit or offset.\n\n— Codex

File: crates/reinhardt-db/src/orm/query.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs]: Implemented in 8ed7ba3cf1. Primary-key and unique-field bulk retrieval paths now reject sliced querysets before issuing a query.\n\n— Codex

File: crates/reinhardt-db/README.md [hunk 5]
```diff
@@ -660,6 +660,53 @@ let updated = User::objects()
...
+        Event::field_created_at().ordering(),
+        Event::field_id().ordering(),
+    ])
+    .await?;
+
+let by_id = Event::objects().all().in_bulk([3_i64, 1, 3]).await?;
+let by_slug = Event::objects()
+    .all()
+    .in_bulk_by(Event::unique_slug(), ["launch", "archive"])
+    .await?;
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/README.md:839]: Implemented in 8ed7ba3cf1. The README example now passes owned String keys, matching the typed unique-field API.\n\n— Codex

File: crates/reinhardt-db/src/orm/query.rs [hunk 6]
```diff
@@ -1627,10 +1631,17 @@ where
...
 			from_alias: Some(alias.to_string()),
+			empty: false,
 			from_subquery_sql: Some(subquery_sql),
 		})
 	}
 
+	/// Marks this queryset as empty without resolving a database connection.
+	pub fn none(mut self) -> Self {
+		self.empty = true;
+		self
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs]: Implemented in 8ed7ba3cf1. update_fields and update_fields_with_conn now return zero immediately for none(), without resolving or invoking an executor.\n\n— Codex

File: crates/reinhardt-core/macros/src/model_derive.rs [hunk 7]
```diff
@@ -3274,6 +3289,61 @@ fn generate_model_form_support(
...
+				.iter()
+				.find(|field| field.name == rust_field_name)
+				.ok_or_else(|| {
+					syn::Error::new_spanned(
+						struct_name,
+						format!("get_latest_by references unknown field '{field_name}'"),
+					)
+				})?;
+
+			if field.rel.is_some() || is_relationship_field_type(&field.ty) {
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-core/macros/src/model_derive.rs:3328]: Implemented compile-time rejection of `#[field(skip = true)]` metadata in `crates/reinhardt-core/macros/src/model_derive.rs`, preventing `get_latest_by` from selecting a non-persisted column.

— Codex

File: crates/reinhardt-db/README.md [hunk 8]
```diff
@@ -660,6 +660,56 @@ let updated = User::objects()
...
+
+let empty = Event::objects().all().none();
+assert_eq!(empty.count().await?, 0);
+```
+
+Unlike Django's string field names, explicit ordering and unique-field bulk
+lookups accept only generated typed field proofs for the same model. Bulk
+retrieval returns a `BTreeMap`, so iteration is sorted by key; duplicate input
+keys collapse and missing keys are omitted. Empty input and `none()` are lazy:
+they do not resolve a connection or invoke an executor. The equivalent
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/README.md:849]: Implemented `none()` short-circuits for `rows_with_db`, `get_composite`, and `get_composite_with_db` in `crates/reinhardt-db/src/orm/query.rs`, before executor or connection access.

— Codex

File: crates/reinhardt-db/src/orm/query.rs [hunk 9]
```diff
@@ -5946,10 +5965,541 @@ where
...
+		ordering: Vec<String>,
+	) -> reinhardt_core::exception::Result<T>
+	where
+		T: serde::de::DeserializeOwned,
+		E: OrmExecutor,
+	{
+		self.ensure_unsliced_retrieval()?;
+		let mut queryset = self.clone();
+		queryset.order_by_fields = ordering;
+		queryset.limit = Some(1);
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:7316]: Cleared `order_by_expressions` whenever the retrieval helpers install their own latest/earliest ordering in `crates/reinhardt-db/src/orm/query.rs`.

— Codex

File: crates/reinhardt-db/src/orm/query.rs [hunk 10]
```diff
@@ -5946,10 +5965,541 @@ where
...
+				"QuerySet retrieval requires at least one typed ordering field".to_string(),
+			));
+		}
+		Ok(fields
+			.iter()
+			.map(|field| {
+				if latest {
+					format!("-{}", field.name())
+				} else {
+					field.name().to_owned()
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:7297]: Documented the nullable-ordering exception and the portable filtering requirement in `crates/reinhardt-db/README.md`.

— Codex

File: crates/reinhardt-db/src/orm/query.rs [hunk 11]
```diff
@@ -5946,10 +5965,541 @@ where
...
+	}
+
+	fn bulk_key_batches<K>(
+		keys: BTreeSet<K>,
+		backend: super::connection::DatabaseBackend,
+	) -> Vec<Vec<K>> {
+		let limit = match backend {
+			super::connection::DatabaseBackend::Sqlite => 900,
+			super::connection::DatabaseBackend::Postgres
+			| super::connection::DatabaseBackend::MySql => 65_535,
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs]: Implemented exact reserved-bind accounting from the cloned queryset SQL before key batching in `crates/reinhardt-db/src/orm/query.rs`. Every DB and executor bulk path now subtracts existing binds from its backend parameter limit.

— Codex

File: crates/reinhardt-db/src/orm/expressions.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/expressions.rs]: Implemented in crates/reinhardt-db/src/orm/expressions.rs: FieldRef::ordering now requires DatabaseField, so only persisted scalar fields can construct OrderingField proofs. Virtual ForeignKeyField and ManyToManyField accessors remain usable for their relation APIs but cannot be used as SQL ordering columns. Added a compile-fail UI case at tests/integration/tests/orm/ui/fail/queryset_relationship_ordering.rs.\n\n---\nCodex attribution: addressed in b091c45c9a.

File: crates/reinhardt-core/macros/src/model_derive.rs [hunk 12]
```diff
@@ -3274,6 +3289,68 @@ fn generate_model_form_support(
...
+					format!("get_latest_by cannot include relation field '{field_name}'")
+				};
+				return Err(syn::Error::new_spanned(struct_name, message));
+			}
+
+			let column = field
+				.config
+				.db_column
+				.clone()
+				.unwrap_or_else(|| rust_field_name.to_owned());
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-core/macros/src/model_derive.rs:3359]: Updated `crates/reinhardt-core/macros/src/model_derive.rs` so generated relation-ID fields resolve `get_latest_by` through the relation metadata when the relation specifies `db_column`.

---
*Resolved by Codex.*

File: crates/reinhardt-db/src/orm/query.rs [hunk 13]
```diff
@@ -5946,10 +5965,561 @@ where
...
+	where
+		T: serde::de::DeserializeOwned,
+	{
+		let ordering = self.metadata_retrieval_ordering(true)?;
+		if self.empty {
+			return Err(Error::NotFound(
+				"No record found matching the query".to_string(),
+			));
+		}
+		let mut conn = super::manager::get_connection().await?;
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:7097]: Updated the global `latest`, `earliest`, `latest_by`, and `earliest_by` helpers to validate unsliced querysets before acquiring the global connection.

---
*Resolved by Codex.*

File: crates/reinhardt-db/tests/orm_queryset_builder.rs [hunk 14]
```diff
@@ -1424,3 +1435,69 @@ fn test_paginate_page_zero_saturates() {
...
+		.order_by(&["-price"])
+		.limit(3)
+		.to_sql()
+		.expect("empty queryset SQL should compile");
+
+	assert_eq!(before_none, after_none);
+	assert!(
+		before_none.contains("WHERE 1 = 0"),
+		"SQL was: {before_none}"
+	);
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/tests/orm_queryset_builder.rs]: Strengthened the deterministic empty-queryset builder regression to compare the complete generated SQL string.

---
*Resolved by Codex.*

File: tests/integration/tests/orm/queryset_docs_ui.rs [hunk 15]
```diff
@@ -4,8 +4,12 @@
 fn queryset_docs_example_compile_pass() {
 	let t = trybuild::TestCases::new();
 	t.pass("tests/orm/ui/pass/queryset_docs_example.rs");
+	t.pass("tests/orm/ui/pass/queryset_typed_retrieval.rs");
 	t.pass("tests/orm/ui/pass/model_enum_typed_filter.rs");
 	t.pass("tests/orm/ui/pass/legacy_field_codec_compat.rs");
 	t.compile_fail("tests/orm/ui/fail/model_enum_string_filter.rs");
 	t.compile_fail("tests/orm/ui/fail/model_enum_integer_filter.rs");
+	t.compile_fail("tests/orm/ui/fail/queryset_latest_cross_model_field.rs");
+	t.compile_fail("tests/orm/ui/fail/queryset_relationship_ordering.rs");
```
[the remark this answers is not in the document]
Comment (kent8192) [reply tests/integration/tests/orm/queryset_docs_ui.rs:13]: Implemented in aeb431ed4da38e0eac5cbf2a2eb308b852119dd1: added the trybuild compile-fail snapshot for virtual relationship ordering and corrected the fixture attribute order so the snapshot records the intended type-bound failure.

---
*Generated by Codex.*

File: crates/reinhardt-db/src/orm/expressions.rs [hunk 16]
```diff
@@ -846,6 +916,20 @@ impl<M, T> FieldRef<M, T> {
...
 
+impl<M, T: DatabaseField> FieldRef<M, T> {
+	/// Convert this persisted scalar field reference into a type-safe ordering field.
+	///
+	/// Relationship fields are virtual model properties and cannot appear in an
+	/// SQL `ORDER BY` clause. Their generated `FieldRef` accessors therefore do
+	/// not satisfy this scalar-field bound.
+	pub const fn ordering(&self) -> OrderingField<M> {
+		OrderingField {
+			name: self.name,
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/expressions.rs:984]: Implemented in e42abbd3bb050b48c258bceb04d82fa8e69a4e2f: `FieldRef::new` now produces an unverified field reference, while `.ordering()` is available only to generated model-field references. Generated accessors use the documented unsafe constructor internally.

---
*Generated by Codex.*

[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/expressions.rs:984]: Implemented in e42abbd3bb050b48c258bceb04d82fa8e69a4e2f: `#[rel(foreign_key, db_column = ...)]` now propagates its physical column to the generated relation-ID field. The integration regression asserts `field_writer_id().name() == "writer_pk"`.

---
*Generated by Codex.*

File: crates/reinhardt-db/src/orm/query.rs [hunk 17]
```diff
@@ -5946,10 +5965,565 @@ where
...
+			super::connection::DatabaseBackend::Sqlite => 900,
+			super::connection::DatabaseBackend::Postgres
+			| super::connection::DatabaseBackend::MySql => 65_535,
+		};
+		let limit = parameter_limit.saturating_sub(reserved_binds).max(1);
+		let mut remaining = keys.into_iter().collect::<Vec<_>>();
+		let mut batches = Vec::new();
+		while !remaining.is_empty() {
+			let count = remaining.len().min(limit);
+			batches.push(remaining.drain(..count).collect());
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs]: Implemented in e42abbd3bb050b48c258bceb04d82fa8e69a4e2f: `bulk_key_batches` now consumes the key set in one linear pass rather than repeatedly draining its front. A SQLite 901-key regression test verifies 900 + 1 batching and preserved order.

---
*Generated by Codex.*

File: crates/reinhardt-db/README.md [hunk 18]
```diff
@@ -660,6 +660,60 @@ let updated = User::objects()
...
+Models can define their default latest/earliest ordering with generated field
+metadata:
+
+```rust
+#[model(
+    app_label = "events",
+    table_name = "events",
+    get_latest_by = ("created_at", "id")
+)]
+struct Event {
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/README.md:815]: Implemented in e42abbd3bb050b48c258bceb04d82fa8e69a4e2f: the typed retrieval README example now imports serde and explicitly derives `Serialize` and `Deserialize` for `Event`.

---
*Generated by Codex.*

File: crates/reinhardt-core/macros/src/model_derive.rs [hunk 19]
```diff
@@ -3274,6 +3289,81 @@ fn generate_model_form_support(
...
+
+			let column = field
+				.config
+				.db_column
+				.clone()
+				.or_else(|| {
+					field
+						.is_fk_id_field
+						.then(|| {
+							let relation_name = rust_field_name.trim_end_matches("_id");
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-core/macros/src/model_derive.rs]: Implemented in 9142f14e2452c1e55967a975c4d235d586526867: relation-ID resolution now uses `strip_suffix("_id")`, so `user_id_id` resolves exactly once to `user_id`. A macro regression verifies that `get_latest_by` selects `user_fk`.

---
*Generated by Codex.*

File: tests/integration/tests/orm/ui/pass/queryset_typed_retrieval.rs [hunk 20]
```diff
@@ -0,0 +1,34 @@
+#![allow(unexpected_cfgs)]
```
[the remark this answers is not in the document]
Comment (kent8192) [reply tests/integration/tests/orm/ui/pass/queryset_typed_retrieval.rs:2]: Implemented in 9142f14e2452c1e55967a975c4d235d586526867: each new trybuild retrieval fixture now explains that the suppression is limited to native-only cfgs emitted by the model macro in its standalone crate.

---
*Generated by Codex.*

File: crates/reinhardt-db/src/orm/query.rs [hunk 21]
```diff
@@ -5946,10 +5965,565 @@ where
...
+		let mut result = BTreeMap::new();
+		for keys in Self::bulk_key_batches(keys, conn.backend(), reserved_binds) {
+			let rows = self
+				.clone()
+				.filter(field.is_in(keys))
+				.all_with_db(conn)
+				.await?;
+			result.extend(
+				rows.into_iter()
+					.filter_map(|model| model.primary_key().map(|key| (key, model))),
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:7597]: Implemented in 9142f14e2452c1e55967a975c4d235d586526867: bulk retrieval restores its lookup column to `values`/`only` projections and removes it from `defer` exclusions. SQLite integration coverage verifies primary-key and unique-key lookup results.

---
*Generated by Codex.*

File: crates/reinhardt-db/src/orm/expressions.rs [hunk 22]
```diff
@@ -114,10 +114,49 @@ impl fmt::Display for F {
...
+/// Such references remain useful for dynamically composed filters, but they
+/// cannot be promoted to [`OrderingField`] without the model macro's proof.
+#[doc(hidden)]
+#[derive(Debug, Clone, Copy)]
+pub enum UnverifiedModelField {}
+
 #[derive(Debug, Clone, Copy)]
-pub struct FieldRef<M, T> {
+/// A typed model field reference whose origin controls ordering eligibility.
+pub struct FieldRef<M, T, Origin = GeneratedModelField> {
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/expressions.rs]: Implemented in 93fe9c325aac66aaa478b41185eb965a9e7f52d8: the default `FieldRef` origin remains generated for source compatibility, while the historical two-parameter `FieldRef::<M, T>::new(...)` constructor explicitly returns an unverified reference. Generated accessors name the generated origin explicitly.

---
*Generated by Codex.*

File: crates/reinhardt-db/src/orm/query.rs [hunk 23]
```diff
@@ -5946,10 +5965,578 @@ where
...
+		conn: &mut E,
+		ordering: Vec<String>,
+	) -> reinhardt_core::exception::Result<T>
+	where
+		T: serde::de::DeserializeOwned,
+		E: OrmExecutor,
+	{
+		self.ensure_unsliced_retrieval()?;
+		let mut queryset = self.clone();
+		queryset.order_by_fields = ordering;
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:7314]: Implemented in 93fe9c325aac66aaa478b41185eb965a9e7f52d8: root column references now qualify whenever legacy or typed select-related joins are present. Added a regression that verifies `ORDER BY "test_users"."id" ASC`.

---
*Generated by Codex.*

File: crates/reinhardt-db/src/orm/expressions.rs [hunk 24]
```diff
@@ -174,15 +262,52 @@ impl<M, T> FieldRef<M, T> {
...
 	}
+}
+
+impl<M, T> FieldRef<M, T, GeneratedModelField> {
+	/// Create an unverified field reference for a dynamically composed filter.
+	///
+	/// This preserves the historical two-parameter `FieldRef::<M, T>::new`
+	/// call shape while preventing dynamic field names from becoming ordering
+	/// proofs.
+	pub const fn new(name: &'static str) -> FieldRef<M, T, UnverifiedModelField> {
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/expressions.rs]: Implemented in 9452473eb6: FieldRef now defaults to the unverified origin, and generated accessors explicitly retain GeneratedModelField proof.

---
*Generated by Codex.*

File: crates/reinhardt-db/src/orm/query.rs [hunk 25]
```diff
@@ -3596,7 +3607,7 @@ where
 	}
 
 	fn root_column_reference(&self, field: &str) -> ColumnRef {
-		if !self.relation_joins.is_empty() && !field.contains('.') {
+		if (!self.relation_joins.is_empty() || self.has_select_related()) && !field.contains('.') {
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs]: Implemented in 9452473eb6: root columns are qualified whenever manual joins are present, including retrieval ordering.

---
*Generated by Codex.*

File: crates/reinhardt-db/src/orm/query.rs [hunk 26]
```diff
@@ -5946,10 +5965,607 @@ where
...
+		if let Some(fields) = &mut self.selected_fields {
+			if !fields
+				.iter()
+				.any(|field| Self::projection_includes_column(field, column))
+			{
+				fields.push(column.to_string());
+			}
+		} else {
+			self.deferred_fields
+				.retain(|field| !Self::projection_includes_column(field, column));
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs]: Implemented in 9452473eb6: bulk lookups resolve physical lookup columns back to logical metadata names before restoring deferred fields.

---
*Generated by Codex.*

File: crates/reinhardt-core/macros/src/model_derive.rs [hunk 27]
```diff
@@ -3345,8 +3441,9 @@ fn generate_field_accessors(
 				///
 				/// Returns a `FieldRef<#struct_name, #field_type>` that provides compile-time
 				/// type safety for field operations.
-				pub const fn #method_name() -> #orm_crate::expressions::FieldRef<#struct_name, #field_type> {
-					#orm_crate::expressions::FieldRef::new(#column_name)
+				pub const fn #method_name() -> #orm_crate::expressions::FieldRef<#struct_name, #field_type, #orm_crate::expressions::GeneratedModelField> {
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-core/macros/src/model_derive.rs]: Implemented in 03f642506c: generated accessors again use the compatible two-parameter `FieldRef<M, T>` defaulting to `GeneratedModelField`, while dynamic `FieldRef::new` retains an explicit unverified return type and cannot create an ordering proof.

---
*Generated by Codex.*

File: crates/reinhardt-db/src/orm/expressions.rs [hunk 28]
```diff
@@ -174,15 +262,52 @@ impl<M, T> FieldRef<M, T> {
...
 	}
+}
 
+impl<M, T> FieldRef<M, T, GeneratedModelField> {
+	/// Create an unverified field reference for a dynamically composed filter.
+	///
+	/// This preserves the historical two-parameter `FieldRef::<M, T>::new`
+	/// call shape while preventing dynamic field names from becoming ordering
+	/// proofs.
+	pub const fn new(name: &'static str) -> FieldRef<M, T, UnverifiedModelField> {
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/expressions.rs]: Implemented in `6b5c85915f9f142136e3efbbc21dd8662af704c7`: `FieldRef<M, T>` again defaults to the compatible unverified type, and `FieldRef::new` returns that exact default. Generated persisted scalar ordering proofs are now emitted separately as `ordering_<field>()` accessors in `crates/reinhardt-core/macros/src/model_derive.rs`, so dynamic field names cannot manufacture ordering proofs.

🤖 Generated with [Codex](https://openai.com/codex)

File: crates/reinhardt-db/src/orm/query.rs [hunk 29]
```diff
@@ -5946,10 +5967,615 @@ where
...
+			if !fields
+				.iter()
+				.any(|field| Self::projection_includes_column(field, column))
+			{
+				fields.push(column.to_string());
+			}
+		} else {
+			let field_name = Self::logical_field_name_for_column(column);
+			self.deferred_fields
+				.retain(|field| !Self::projection_includes_column(field, &field_name));
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:7410]: Implemented in `6b5c85915f9f142136e3efbbc21dd8662af704c7`: deferred projections now translate retained model metadata through `db_column_name()` in `crates/reinhardt-db/src/orm/query.rs`, including the multi-field `defer` path. The regression covers `defer(&["id", "title"]).in_bulk([1])` with `id` mapped to `article_id`.

🤖 Generated with [Codex](https://openai.com/codex)

File: crates/reinhardt-db/src/orm/query.rs [hunk 30]
```diff
@@ -3596,7 +3607,9 @@ where
 	}
 
 	fn root_column_reference(&self, field: &str) -> ColumnRef {
-		if !self.relation_joins.is_empty() && !field.contains('.') {
+		if (!self.relation_joins.is_empty() || !self.joins.is_empty() || self.has_select_related())
+			&& !field.contains('.')
+		{
 			ColumnRef::table_column(Alias::new(self.root_alias()), Alias::new(field))
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:4594]: Implemented in `6b5c85915f9f142136e3efbbc21dd8662af704c7`: `root_column_reference` now exempts known annotation and selected-expression aliases from root qualification across joins. A regression in `crates/reinhardt-db/src/orm/query.rs` verifies that manual joins retain alias-based `ORDER BY "other_age" ASC`.

🤖 Generated with [Codex](https://openai.com/codex)

File: crates/reinhardt-db/src/orm/query.rs [hunk 31]
```diff
@@ -7067,6 +7082,607 @@ where
...
+		keys: BTreeSet<K>,
+		backend: super::connection::DatabaseBackend,
+		reserved_binds: usize,
+	) -> Vec<Vec<K>> {
+		let parameter_limit: usize = match backend {
+			super::connection::DatabaseBackend::Sqlite => 900,
+			super::connection::DatabaseBackend::Postgres
+			| super::connection::DatabaseBackend::MySql => 65_535,
+		};
+		let limit = parameter_limit.saturating_sub(reserved_binds).max(1);
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs]: Implemented in `46bf1d4e2a9bc8c07e8606bedb2c6e8030e6ba30`: `bulk_key_batches` in `crates/reinhardt-db/src/orm/query.rs` now returns a validation error for nonempty lookups when `reserved_binds >= parameter_limit`, and all primary-key/unique-field connection and executor paths propagate that error. The regression covers both PostgreSQL and MySQL at 65,535 reserved binds, and the README documents the source-query bind budget.

🤖 Generated with [Codex](https://openai.com/codex)

File: crates/reinhardt-db/src/orm/query.rs [hunk 32]
```diff
@@ -12540,6 +13285,43 @@ mod tests {
...
 
+	#[test]
+	fn test_manual_join_qualifies_root_ordering_columns() {
+		let sql = QuerySet::<TestUser>::new()
+			.inner_join_on::<TestProject>("test_users.id = test_projects.user_id")
+			.order_by(&["id"])
+			.to_sql()
+			.expect("query SQL should compile");
+
+		assert!(sql.contains(r#"ORDER BY "test_users"."id" ASC"#));
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:13326]: Implemented in `46bf1d4e2a9bc8c07e8606bedb2c6e8030e6ba30`: `test_manual_join_preserves_annotation_ordering_aliases` now uses `assert_eq!` against the complete generated SQL, including projection, annotation, join, and the unqualified ordering alias.

🤖 Generated with [Codex](https://openai.com/codex)
